staging/src/k8s.io/kube-aggregator/pkg/controllers/status/metrics.go

```diff
@@ -38,17 +44,65 @@ var (
 		},
 		[]string{"name", "reason"},
 	)
-	unavailableGauge = metrics.NewGaugeVec(
-		&metrics.GaugeOpts{
-			Name:           "aggregator_unavailable_apiservice",
-			Help:           "Gauge of APIServices which are marked as unavailable broken down by APIService name.",
-			StabilityLevel: metrics.ALPHA,
-		},
+
+	unavailableGaugeDesc = metrics.NewDesc(
+		"aggregator_unavailable_apiservice",
+		"Gauge of APIServices which are marked as unavailable broken down by APIService name.",
 		[]string{"name"},
+		nil,
+		metrics.ALPHA,
+		"",
 	)
 )
+var registerMetrics sync.Once
+
+// Register registers metrics for status controller.
+func Register(apiServiceLister listers.APIServiceLister) {
+	registerMetrics.Do(func() {
+		legacyregistry.CustomMustRegister(newAPIServiceStatusCollector(apiServiceLister))
+		legacyregistry.MustRegister(unavailableCounter)
+	})
+}
+
+type apiServiceStatusCollector struct {
+	metrics.BaseStableCollector
+
+	apiServiceLister listers.APIServiceLister
+}
+
+// Check if apiServiceStatusCollector implements necessary interface.
+var _ metrics.StableCollector = &apiServiceStatusCollector{}
+
+func newAPIServiceStatusCollector(apiServiceLister listers.APIServiceLister) *apiServiceStatusCollector {
+	return &apiServiceStatusCollector{
+		apiServiceLister: apiServiceLister,
+	}
+}
+
+// DescribeWithStability implements the metrics.StableCollector interface.
+func (c *apiServiceStatusCollector) DescribeWithStability(ch chan<- *metrics.Desc) {
+	ch <- unavailableGaugeDesc
+}
 
-func init() {
-	legacyregistry.MustRegister(unavailableCounter)
-	legacyregistry.MustRegister(unavailableGauge)
+// CollectWithStability implements the metrics.StableCollector interface.
+func (c *apiServiceStatusCollector) CollectWithStability(ch chan<- metrics.Metric) {
+	apiServiceList, _ := c.apiServiceLister.List(labels.Everything())
+	for _, apiService := range apiServiceList {
```

Comment:
the more I think about this pattern, the more I dislike it. IMO it inverses direction in the wrong way: metrics are global (they shouldn't be, but it's where we are), and leaking a client into a global construct makes this wrongly shaped construct even bigger. Instead we should strive to remove globals and any logic connected to them.

Reply:
when I was writing my first draft for this PR, I took a different approach that is quite different from what's done in other packages but considering your comment, it might be better suited. The custom collector had a map, that was updated by the controller with the availabilities of the apiservices. Upon collection, the metrics were created based on the information stored in the map instead of the ones from the lister.